### PR TITLE
Show layer status for all visible layers

### LINKED ISSUE #2955: Feature: show intensity value in grid mode
## 🚀 Feature
Not sure how hard this would be, but it would be nice to be able to get the intensity value of _any_ pixel when hovering in grid mode.  Currently: you will only see the pixel value if you happen to hover over the selected layer in the grid... but since there's no overlapping layers in grid mode, it should in theory be possible for all layers that support `get_value`?

### PR DESCRIPTION
**Pull request by jni**
Closes #2955!

This one was a long time coming!

Rather than only showing the status for the active layer, we now show
the status for all visible and loaded layers. The active layer is
indicated by `###`.

I initially skipped all layers behind a translucent layer with opacity=1.
However, when I was testing it with `example/add_points.py`, it felt
quite jarring to have the image layer disappear from the status just as
we hovered over the point. Sure, it could be remedied by setting opacity
to less than 1, but I thought it's better to just show them all. Happy
to revisit this decision if folks play around with it and find they'd
rather not see hidden layers.

This also allows hovering over *all* layers in a grid mode and seeing the
right layer info on the status bar, without changing the selection. It's
very nice!

Will upload short video capture shortly.

### CODE REVIEW

**Comment by jni** [comment]
Video capture:

https://github.com/user-attachments/assets/ea8c61c6-8c21-4467-84aa-1ea5c5a840c5

**Comment by jni** [comment]
Points layer with properties:

https://github.com/user-attachments/assets/51334413-156e-48cc-9ab4-5d4d2805b1a4

**Comment by jni** [comment]
Tests are evidently borked, not too surprising. 😅 Will check them out in the morning, but in the meantime, definitely appreciate feedback about the functionality! btw, in case you are tempted to ask: using the QStatusBar we are limited to plaintext: no bold, no (partial) colours. It would be nice to bold or otherwise highlight the active layer, but that's not easy, I'd rather leave that for the (far) future.

**Review by Czaki — CHANGES_REQUESTED** [review]
Hm. I have artificial labels with huge properties and even when I select 'nuclei' layer I do not see the value of voxel below the mouse. 
So we need to find a way to ensure that data, from the current selected layer, are visible. 

At second. `>>>` `//` and `###` look strange to me. I prefer to look for some better looking unicode character. 

From some reason, the tooltip is not showin, but I need to debug this. 

I will try to review more later. 


Screenshot 2025-03-06 at 14 47 20

**Comment by jni** [comment]
> Hm. I have artificial labels with huge properties and even when I select 'nuclei' layer I do not see the value of voxel below the mouse.

Yeah, I noticed the properties could be an issue. I don't know the best way to fix that but "huge properties" are a problem for the status bar no matter what — I was thinking that maybe we leave that to the tooltip only. (The tooltip still only works for active layer.)

> At second. >>> // and ### look strange to me. I prefer to look for some better looking unicode character.

Yup. I actually like the >>> but in general I would very much be open towards improving these markers. Suggestions are very welcome!

File: napari/components/viewer_model.py [hunk 1]
```diff
@@ -580,28 +580,55 @@ def _calc_status_from_cursor(
...
                 world=True,
             )
-
-            if self.tooltip.visible:
-                tooltip_text = active._get_tooltip_text(
-                    np.asarray(self.cursor.position),
-                    view_direction=np.asarray(self.cursor._view_direction),
-                    dims_displayed=list(self.dims.displayed),
-                    world=True,
+            emphasis = ' ### ' if layer is active else ''
```
Comment (TimMonko) [inline napari/components/viewer_model.py]: Much smaller, and looks nice and set out still in my eye
```suggestion
            emphasis = ' ** ' if layer is active else ''
```

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -580,28 +580,55 @@ def _calc_status_from_cursor(
...
-            else:
-                tooltip_text = ''
+        if coord2val:
+            if not self.grid.enabled:
+                # use a single coordinate system
+                values = list(itertools.chain(*coord2val.values()))
+                key = next(iter(coord2val))  # choose arbitrary coordinate
+                coord2val = {key: values}
+            status_strs = [
+                key + ' // '.join(values) for key, values in coord2val.items()
```
Comment (TimMonko) [inline napari/components/viewer_model.py]:
```suggestion
                key + ' || '.join(values) for key, values in coord2val.items()
```

**Comment by psobolewskiPhD** [comment]
Just going to link a related issue: https://github.com/napari/napari/issues/2253

File: napari/components/viewer_model.py (hunk 1 above)
Comment (jni) [reply napari/components/viewer_model.py]: I actually had the stars before but my eyes struggled to find them...

**Comment by jni** [comment]
Noting down some discussion from the [2025-03-06/07 Pacific community meeting](https://hackmd.io/VSaKh4hMQ12Bn7uWD_ij6A#2025-03-0607):

- having all visible layers is too disruptive because:
  - if there's lots of layers or a layer's info is too long, there is now no easy way to inspect the selected layer. (Other than making all other layers invisible.)
  - the position of the status value is much more variable with respect to what you are looking for
- instead, we can make this PR strictly an improvement by:
  - only using all visible layers for grid mode (which is currently broken) (ie fix #2955)
  - in non-grid mode, using *selected* layers. This is a strict improvement because currently when there is more than one layer selected, you get the useless "Ready" message.
- in the future, we can change the behaviour more broadly behind an opt-in preference/feature flag, if folks are clamouring for it.

@Czaki also pointed out that we could do better status updates if we pass more information in the status dictionary and dynamically change the status based on preferences and layer properties. Just for my own record, the place to change this is here:

https://github.com/napari/napari/blob/f5de441ebf47960fa0a626cb1688e143a6412723/napari/_qt/qt_main_window.py#L1586-L1605

and here:

https://github.com/napari/napari/blob/838ce4cc6b5cd77a5197102ee7e69e57d7a0358d/napari/_qt/widgets/qt_viewer_status_bar.py#L72-L96

The other thing I wanted to write down while testing this is that I think the debounce on canvas tooltips is too slow, and persists too long after moving — actually I think moving the mouse *at all* should dismiss the tooltip. Will try to remember to make an issue about it.

File: napari/components/viewer_model.py (hunk 2 above)
Comment (jni) [reply napari/components/viewer_model.py]: 😂 

Seriously, I'm open to changing it, but after some consensus — personally, I prefer the slashes. 🙏

**Comment by jni** [comment]
Lookin' good...!

CleanShot 2025-03-07 at 16 49 15@2x

I still need to fix tests, but otherwise I think this is ready for a proper kick of the tires!

- @Czaki I think that switching to only selected layers solves your objection?
- @psobolewskiPhD similarly
- @TimMonko and everyone: key to resolve now is the separation markers. I don't have strong opinions here, but here's the current state of play:
  - ` ### ` for the active layer. @TimMonko [suggests](https://github.com/napari/napari/pull/7673#discussion_r1984137085) ` ** `, which I like less — I did try it and found it hard to find the right layer. *However*, with multiple layers selected, there is no active layer, so actually I think we can just remove this altogether?
  - ` // ` to separate the layers. @TimMonko [suggests](https://github.com/napari/napari/pull/7673#discussion_r1984138063) ` || `, which I also like less, but 🤷 happy to go with overall consensus.
  - ` ||| ` to separate "groups" of layers in grid mode when stride is >=2. I think that this separator should be visually distinct from the layers separator, so if we swap that one to ` || `, then I would want a different glyph here. But again, happy to look for consensus.

**Comment by Czaki** [comment]
Maybe `»` instead of `>>`
We do not have an active layer now. 
Maybe `⬦` instead of `||`
Maybe `⬥` instead of `|||`

Why you expect report value from more than one layer in grid mode?

**Comment by jni** [comment]
I love the symbol suggestions!

> Why do you expect report value from more than one layer in grid mode?

In grid mode you can set stride >1 and then you get layers gridded in groups. So eg if you have a bunch of images interleaved with their segmentations in the layer list you can see each image with its segmentation laid out in a grid.

**Review by TimMonko — COMMENTED** [review]
Overall, this implementation is really really clever and an absolute all around improvement to functionality for me. For many users, napari is _the source of understanding_ of image data. When teaching prior to this PR, its constantly an issue to be clicking back to activate a layer to show someone the value under the cursor on different images. This PRs behavior makes it so that multi-channel images are much easier to explore and understand via selecting multiple layers. It also makes it much easier to understand how various convolutions, etc, change underlying image values because now _both_ the original and any processed images can be investigated at the same time. This is HUGE.

I investigated its functionality in grid mode, with various strides (another reason for the multi-group @Czaki) and spacings, and all works as expected. 

I also like the newly suggested characters, really for me anything that is obvious but not too wide will make me happy. 

I'll try to carefully check over the implementation later, but I could not find any bugged edge cases personally. (aside from an already present bug in #7585)

For me, I think it would be nicer if the spacers were added a bit farther up as constants in the function, just to make finding them easier. 

Some current issues still present, but are not blockers for me:
1. Really long layer names clutter up the status (and does currently). Perhaps elide to a certain constant length, like 10? Layer names are elided but dependent on the width of the sidebar. 
2. This PR does not aim to fix #7585, where 2D layers in 3D dont show the correct positions. I wonder if that would be hard to implement here? I don't have a ton of time for my review atm. 

Possible alternative implementation:
1.  Use the layer index (#7675 needs implemented) instead of layer name in status bar. This would keep the status much shorter, especially in cases of many layers or long layer names.

**Comment by psobolewskiPhD** [comment]
First, I do like the feature part, being able to see values of multiple layers.
But I think more thought needs to go into the polish, because the status bar and getting values is such a high visual traffic area, so it's important to get it right or as close to right as possible.

Worth keeping in mind, that one important thing that is lost here is the filename in the status bar (and the plugin that loaded it, but that IMO is less important?). I think having the filename shown is a very useful thing; layer names can be generic things like channel names or whatever, not indicating what data file one is working with.

Other thoughts:
- I can't figure out what `>>>` is meant to indicate.
- I'm not sure I would understand what the `###` is if it wasn't for reading through this thread.
If it's grid-mode only, then I think it should only be used there. It's a lot of clutter, why is it duplicated? Can we do rich text in the status? because then **bold** for the selected layer(s) would be better?
- I dislike the `//` as field separator; visually it looks odd and I've not seen something like that for separating fields. I would suggest a simpler, vertical element, like `|` -- also used for cells in markdown -- maybe with extra spaces around it?
- I really dislike how skittish and glitchy the status bar seems now with the text wiggling around as you mouse over, even with a single layer showing and selected, but it's way worse with more layers or longer names. There's always been a bit of wiggle, because the coords and value would change spacing, but the status bar was "anchored" by the static information. Now, it's way worse, because nothing is static: when the values or number spacing changes, everything in the the status ends up moving around. Are we able to use a monospace font for the status bar?

**Comment by jni** [comment]
@TimMonko

> Perhaps elide to a certain constant length, like 10? 

I think this is an existing issue, so it could/should be fixed in an independent PR.

> where 2D layers in 3D dont show the correct positions. I wonder if that would be hard to implement here? 

Yes, it would. 😂 Totally different part of the codebase, where 3D coordinates get turned into data coordinates. So it should be a different PR.

> Use the layer index

Once we display layer indices on the layer list, this would be a good idea.

@psobolewskiPhD 

> I can't figure out what >>> is meant to indicate.

It's like a `:` but bigger 😂. What do you think about @Czaki's [suggestion](https://github.com/napari/napari/pull/7673#issuecomment-2705974272) of `»`?

> I'm not sure I would understand what the ### is if it wasn't for reading through this thread.

That's obsolete now that we only show multiple statuses when there's a multi-selection. Will remove.

> I dislike the // as field separator;

See @Czaki's suggestion of `⬦` above. Thoughts? (I did initially try single characters, `/` or `|`, but they were really hard to make out.)

> Can we do rich text in the status?

We cannot, see [this comment](https://github.com/napari/napari/pull/7673#issuecomment-2703762091). I immediately reached for bold (and layer colormaps) but it's not a thing. 😢 I think in the future we could make our own statusbar widget that has more functionality.

> Are we able to use a monospace font for the status bar?

I think QSS should let us do this, and indeed I've been thinking about this ever since we have a status bar. I don't really understand why you think it's worse now though? Whether it's "anchored" by the static information or by the left edge of the bar, it's the same amount of movement?

The static information (filename+plugin) comes from #4664, which I missed at the time but @Czaki [opposed](https://github.com/napari/napari/pull/4664#pullrequestreview-1022528679) having so much info in the status bar (when it can be and is had in the layer tooltip), and I agree with him. However, we can reserve this discussion for a later issue/PR and I can special case the selection vs active so that the active layer status doesn't change, and this PR strictly affects only grid mode and multi-layer selection.

**Comment by psobolewskiPhD** [comment]
> It's like a : but bigger 😂. What do you think about @Czaki's https://github.com/napari/napari/pull/7673#issuecomment-2705974272 of »?

But what is it supposed to mean or indicate? I'm sorry to be dense, I just don't get it.

Re: field separators: they should be "hard to make out" -- they should serve the job and blend in. You shouldn't be aware of them if they work properly. Do you notice the lines between Excel table elements? You want to be able to focus on the important info, so I'm wary of the special symbol characters, they are atypical so they tend to draw the eye and attention, thus working in the opposite way of what we want.

> Whether it's "anchored" by the static information or by the left edge of the bar, it's the same amount of movement?

With the things I tried, it's not the same visually.
On live it's just the coordinate and value and wiggle -- on the far right.
With this PR, everything wiggles almost all the time, because of the way the coordinates are on the far left -- and everything works off that -- and then values are intermixed with the names.

**Comment by jni** [comment]
> But what is it supposed to mean or indicate? I'm sorry to be dense, I just don't get it.

I dunno, an arrow? These coordinates give this value. These coordinates correspond to this value. We are at these coordinates, therefore -> we see this value.? What does the colon indicate?

> Do you notice the lines between Excel table elements?

No, but they are perfectly lined up — and I do notice them if the values are bigger than the cell, so there is a tradeoff. I think a more appropriate analogy is how easy it is to find the 4th column in a .csv file — not at all, and we would prefer a stronger visual delimiter between the values if the task was to find the nth value visually quickly.

**Comment by Czaki** [comment]
> But what is it supposed to mean or indicate? I'm sorry to be dense, I just don't get it.

I suggest replacing `>>` with `»` because it is one character. Also, for others, I try to find distinguish symbols that occupy single character position to make text more compact and have more information visible on small screens.

More ideas:

1) do not show properties of layer on the status bar if more than one layer is selected. It could be represented still when using a tooltip. 
2) Show help text when no layer is selected to inform about the new behavior.

**Comment by melonora** [comment]
Great work @jni! Some comments on what has been said before,

- Possible alternative implementation using index of layer in layer list as opposed to name #7675, personally I would vote against this as it does require a double lookup.

That being said long layer names are making it quite cluttered, perhaps the only solution there is a floating widget for status of layers, but not certain how all of you think of that.

Regarding separators, I find `|` to be more natural than the other special characters, though I understand it can be difficult to make out. However, if long term we think this would be the best (and perhaps make it better visible later with our own status bar) could we already live with it now instead of having to change characters once we would move to a new implementation?

Also regarding @psobolewskiPhD point regarding file names. I agree that this is very informative and it is information that a lot of people also want. That being said I think this is more metadata where viewers like FIJI have a separate floating window for metadata. Would it be an idea when right clicking a layer in the layer list to have `layer info` which would open a floating window with basic layer metadata? This could even contain information like the resolution etc.

In any case, lets discuss a bit further during the next community meeting to agree on some point to get this PR over the line.

**Comment by willingc** [comment]
Another approach that could be simpler would be to use a separator of 4 spaces between layers in the status bar. `separator = '    '` could be set at the beginning of the status bar function and used in place of the strings with spacers. It's visually clean. See below.

Screenshot 2025-03-11 at 3 54 05 PM

**Comment by Czaki** [comment]
I do not like `|` as a separator. I personably treat it as a part of the text and need to parse text to spot parts of it.

**Comment by willingc** [comment]
Let's move closer to a **decision on this PR** so we can merge before Friday's alpha1 release.

> Maybe » instead of >>
We do not have an active layer now.
Maybe ⬦ instead of ||
Maybe ⬥ instead of |||

I would be in favor of using this suggestion for the alpha1 release. It can always be adjusted if there is feedback during the 2 alphas. Any objections? 

cc/ @jni @Czaki @psobolewskiPhD @TimMonko @DragaDoncila @melonora

**Comment by TimMonko** [comment]
### Today's community meeting feedback
1. If you ask 100 people, you will get 100 different opinions 
2. Some users preferred unicode characters for spacing out the layer statuses, others preferred Carol's suggestion of tabs. 
3. More information about images _should_ be more easily available in the GUI (e.g. we lose reader info with this PR), _but_ the status bar is **not** the place to make this information available. (I'll open an Issue for this) 

### To Recap this PR
1. **Maintains** current position + intensity functionality for single-layer selections 
2. **Fixes** status bar in grid mode 
3. **Adds** multi-layer status support
4. **Enhances** status bar in grid mode by allowing multiple layer
6. **Removes** reader info from layer information in status bar

### My thoughts 
This is **multiple** improvements, with one small regression (that we can more appropriately address via different means in the GUI). And as such this PR should go through (I would have approved it almost right away for sure, but I figure it would be good to have tests passing). I also greatly appreciate the implementation.

Feedback should be gathered during the alpha about the separators, in case there are any strong opinions or edge cases we didn't think about. These would be easily changed in followup PRs. Other follow ups include how to manage long layer names (again, not a current regression) and potentially how to help @psobolewskiPhD not feel as if the status bar is so twitchy as positions change (but again, this is not a regression from current functionality)

My hope is that this gets wrapped up and merged ASAP. From working with users this represents a big step forward in the usability and teaching opportunity with napari, with extremely limited losses (i.e. only adds visual clutter when going out of way to add it, and loss of reader info).

Also, the entire point of alpha's is for feedback :) We need to keep in mind that better is the enemy of the good.

**Comment by TimMonko** [comment]
p.s, here is my suggestion a unicode to separate position and then spaces to separate layers. here's a 'realistic' situation

[jni/napari#33](https://github.com/jni/napari/pull/33)

image

**Comment by melonora** [comment]
@TimMonko thanks for summarizing and I think the last suggestion looks clean. I would also like to avoid postponing merging of this PR given that it doesn't have regressions. I will fix the tests now

**Comment by willingc** [comment]
Let's implement @TimMonko's suggestion and move forward :heart:

**Review by TimMonko — APPROVED** [review]

**Comment by willingc** [comment]
I updated the PR to main HEAD. Hopefully, that will resolve CircleCI's failures.

File: napari/components/_tests/test_viewer_model.py [hunk 3]
```diff
@@ -1044,32 +1056,14 @@ def test_get_status_text():
...
     viewer.tooltip.visible = False
     assert viewer._calc_status_from_cursor() == (
-        {
-            'coordinates': ' [1 2]: 0; a: 1',
-            'layer_base': 'Labels',
-            'layer_name': 'Labels',
-            'plugin': '',
-            'source_type': '',
-        },
+        ' [1 2] » Labels: 0; a: 1    ',
```
Comment (Czaki) [inline napari/components/_tests/test_viewer_model.py:1059]: I don't like this change. I prefer to return even richer dict and construct the string in the status bar.

**Review by Czaki — APPROVED** [review]
I think that my remarks were addressed. 
I think that we should do in the direction, where we provide rich properties dict and then calculate text in the status bar (based on available space), but we may do this in a follow up PR.

**Comment by jni** [comment]
@willingc and I are gearing up for 0.6.0a0 and we've decided to merge this and iterate for the later alphas. @psobolewskiPhD I think most people like the arrow, so let's give it a go. The monospace is possible but let's aim for that in an upcoming PR. (Still before final release hopefully.)

File: napari/components/viewer_model.py [hunk 4]
```diff
@@ -580,28 +580,63 @@ def _calc_status_from_cursor(
...
+        if self.tooltip.visible and active is not None and active._loaded:
+            tooltip_text = active._get_tooltip_text(
+                np.asarray(self.cursor.position),
+                view_direction=np.asarray(self.cursor._view_direction),
+                dims_displayed=list(self.dims.displayed),
+                world=True,
+            )
 
-        return 'Ready', ''
+        return status_str, tooltip_text
```
Comment (jni) [inline napari/components/viewer_model.py:639]: Here I changed return from dict to a string; it was unintentional to change it always, rather than only in grid and multiple-selection modes. Will update in upcoming PR.
